File: app/src/main/java/com/arjanvlek/oxygenupdater/Support/Dialogs.java
package com.arjanvlek.oxygenupdater.Support;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.StringRes;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;

import com.arjanvlek.oxygenupdater.BuildConfig;
import com.arjanvlek.oxygenupdater.Model.OxygenOTAUpdate;
import com.arjanvlek.oxygenupdater.R;
import com.arjanvlek.oxygenupdater.views.MessageDialog;

public class Dialogs {

    /**
     * Shows an {@link MessageDialog} with the occured download error.
     * @param title Title of the error message
     * @param message Contents of the error message
     */
    public static void showDownloadError(final Fragment fragment, final UpdateDownloader updateDownloader, final OxygenOTAUpdate oxygenOTAUpdate, @StringRes int title, @StringRes int message) {
        showDownloadError(fragment, updateDownloader, oxygenOTAUpdate, fragment.getString(title), fragment.getString(message));
    }

    /**
     * Shows an {@link MessageDialog} with the occured download error.
     * @param title Title of the error message
     * @param message Contents of the error message
     */
    public static void showDownloadError(final Fragment fragment, final UpdateDownloader updateDownloader, final OxygenOTAUpdate oxygenOTAUpdate, String title, String message) {
        MessageDialog errorDialog = new MessageDialog()
                .setTitle(title)
                .setMessage(message)
                .setPositiveButtonText(fragment.getString(R.string.download_error_close))
                .setNegativeButtonText(fragment.getString(R.string.download_error_retry))
                .setClosable(true)
                .setDialogListener(new MessageDialog.DialogListener() {
                    @Override
                    public void onDialogPositiveButtonClick(DialogFragment dialogFragment) {
                        Notifications.hideDownloadCompleteNotification(fragment.getActivity());
                    }

                    @Override
                    public void onDialogNegativeButtonClick(DialogFragment dialogFragment) {
                        Notifications.hideDownloadCompleteNotification(fragment.getActivity());
                        updateDownloader.cancelDownload(oxygenOTAUpdate);
                        updateDownloader.downloadUpdate(oxygenOTAUpdate);
                    }
                });
        errorDialog.setTargetFragment(fragment, 0);
        FragmentTransaction transaction = fragment.getActivity().getSupportFragmentManager().beginTransaction();
        transaction.add(errorDialog, "DownloadError");
        transaction.commitAllowingStateLoss();
    }

    public static void showNoNetworkConnectionError(Fragment fragment) {
        MessageDialog errorDialog = new MessageDialog()
                .setTitle(fragment.getString(R.string.error_app_requires_network_connection))
                .setMessage(fragment.getString(R.string.error_app_requires_network_connection_message))
                .setNegativeButtonText(fragment.getString(R.string.download_error_close))
                .setClosable(false);
        errorDialog.setTargetFragment(fragment, 0);
        errorDialog.show(fragment.getFragmentManager(), "NetworkError");
    }

    public static void showServerMaintenanceError(Fragment fragment) {
        MessageDialog serverMaintenanceErrorFragment = new MessageDialog()
                .setTitle(fragment.getString(R.string.error_maintenance))
                .setMessage(fragment.getString(R.string.error_maintenance_message))
                .setNegativeButtonText(fragment.getString(R.string.download_error_close))
                .setClosable(false);
        serverMaintenanceErrorFragment.setTargetFragment(fragment, 0);
        serverMaintenanceErrorFragment.show(fragment.getFragmentManager(), "MaintenanceError");
    }

    public static void showAppOutdatedError(Fragment fragment) {
        MessageDialog appOutdatedErrorFragment = new MessageDialog()
                .setTitle(fragment.getString(R.string.error_app_outdated))
                .setMessage(fragment.getString(R.string.error_app_outdated_message))
                .setPositiveButtonText(fragment.getString(R.string.error_google_play_button_text))
                .setNegativeButtonText(fragment.getString(R.string.download_error_close))
                .setClosable(false)
                .setDialogListener(new MessageDialog.DialogListener() {
                    @Override
                    public void onDialogPositiveButtonClick(DialogFragment dialogFragment) {
                        try {
                            final String appPackageName = BuildConfig.APPLICATION_ID;
                            try {
                                fragment.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
                            } catch (ActivityNotFoundException e) {
                                fragment.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
                            }
                        } catch (Exception ignored) {

                        }
                    }

                    @Override
                    public void onDialogNegativeButtonClick(DialogFragment dialogFragment) {

                    }
                });
        appOutdatedErrorFragment.setTargetFragment(fragment, 0);
        appOutdatedErrorFragment.show(fragment.getFragmentManager(), "AppOutdatedError");
    }

}

File: app/src/main/java/com/oxygenupdater/dialogs/Dialogs.kt
package com.oxygenupdater.dialogs

import android.app.Activity
import android.content.DialogInterface.BUTTON_NEGATIVE
import android.content.DialogInterface.BUTTON_NEUTRAL
import android.content.DialogInterface.BUTTON_POSITIVE
import androidx.annotation.StringRes
import com.oxygenupdater.R
import com.oxygenupdater.extensions.openPlayStorePage
import com.oxygenupdater.internal.KotlinCallback
import com.oxygenupdater.utils.LocalNotifications

object Dialogs {

    /**
     * Shows a [MessageDialog] with the occurred download error.
     *
     * @param title   Title of the error message
     * @param message Contents of the error message
     */
    fun showDownloadError(
        activity: Activity?,
        isResumable: Boolean,
        @StringRes title: Int,
        @StringRes message: Int,
        callback: KotlinCallback<Boolean>? = null
    ) = checkPreconditions(activity) {
        showDownloadError(
            activity!!,
            isResumable,
            activity.getString(title),
            activity.getString(message),
            callback
        )
    }

    /**
     * Shows a [MessageDialog] with the occurred download error.
     *
     * @param title   Title of the error message
     * @param message Contents of the error message
     */
    fun showDownloadError(
        activity: Activity?,
        isResumable: Boolean,
        title: String?,
        message: CharSequence?,
        callback: KotlinCallback<Boolean>? = null
    ) = checkPreconditions(activity) {
        MessageDialog(
            activity!!,
            title = title,
            message = message,
            positiveButtonText = when {
                callback == null -> null
                isResumable -> activity.getString(R.string.download_error_resume)
                else -> activity.getString(R.string.download_error_retry)
            },
            negativeButtonText = activity.getString(R.string.download_error_close),
            positiveButtonIcon = when {
                callback == null -> null
                isResumable -> R.drawable.download
                else -> R.drawable.auto
            },
            cancellable = true
        ) {
            when (it) {
                BUTTON_POSITIVE -> {
                    LocalNotifications.hideDownloadCompleteNotification()
                    callback?.invoke(isResumable)
                }
                BUTTON_NEGATIVE -> LocalNotifications.hideDownloadCompleteNotification()
                BUTTON_NEUTRAL -> {
                    // no-op
                }
            }
        }.show()
    }

    /**
     * Shows a [MessageDialog] with the occurred article loading error.
     *
     * @param title   Title of the error message
     * @param message Contents of the error message
     */
    fun showArticleError(
        activity: Activity?,
        isRefreshable: Boolean,
        title: CharSequence?,
        message: CharSequence?,
        callback: () -> Unit
    ) = checkPreconditions(activity) {
        MessageDialog(
            activity!!,
            title = title?.toString(),
            message = message,
            positiveButtonText = if (isRefreshable) {
                activity.getString(R.string.download_error_retry)
            } else {
                null
            },
            negativeButtonText = activity.getString(R.string.download_error_close),
            positiveButtonIcon = if (isRefreshable) {
                R.drawable.auto
            } else {
                null
            },
            cancellable = true
        ) {
            when (it) {
                BUTTON_POSITIVE -> if (isRefreshable) {
                    callback.invoke()
                }
            }
        }.show()
    }

    fun showServerMaintenanceError(activity: Activity?) = checkPreconditions(activity) {
        MessageDialog(
            activity!!,
            title = activity.getString(R.string.error_maintenance),
            message = activity.getString(R.string.error_maintenance_message),
            negativeButtonText = activity.getString(R.string.download_error_close),
            cancellable = false
        ).show()
    }

    fun showAppOutdatedError(activity: Activity?) = checkPreconditions(activity) {
        MessageDialog(
            activity!!,
            title = activity.getString(R.string.error_app_outdated),
            message = activity.getString(R.string.error_app_outdated_message),
            positiveButtonText = activity.getString(R.string.error_google_play_button_text),
            negativeButtonText = activity.getString(R.string.download_error_close),
            cancellable = false
        ) {
            when (it) {
                BUTTON_POSITIVE -> activity.openPlayStorePage()
                BUTTON_NEGATIVE -> {
                    // no-op
                }
                BUTTON_NEUTRAL -> {
                    // no-op
                }
            }
        }.show()
    }

    fun showAdvancedModeExplanation(activity: Activity?, callback: KotlinCallback<Boolean>) = checkPreconditions(activity) {
        MessageDialog(
            activity!!,
            title = activity.getString(R.string.settings_advanced_mode),
            message = activity.getString(R.string.settings_advanced_mode_explanation),
            positiveButtonText = activity.getString(R.string.enable),
            negativeButtonText = activity.getString(android.R.string.cancel),
            positiveButtonIcon = R.drawable.done_outline,
            cancellable = true
        ) {
            when (it) {
                BUTTON_POSITIVE -> callback.invoke(true)
                BUTTON_NEGATIVE -> callback.invoke(false)
                BUTTON_NEUTRAL -> {
                    // no-op
                }
            }
        }.show()
    }

    private fun checkPreconditions(activity: Activity?, callback: () -> Unit) {
        if (activity?.isFinishing == false) {
            callback.invoke()
        }
    }
}
